### ISSUE TITLE
💡 RFC: Additional file hashing in production build

### ISSUE THREAD

**Issue by stramel** [snowpackjs/astro#874]
### Background & Motivation

This stems from a conversation on adding Astro to the supported Vercel frameworks. https://github.com/vercel/vercel/pull/6477#discussion_r669925465

Currently, we have some file hashing going on in the `_astro` folder for `.css` files. 

Handling additional assets would be ideal and I believe is intended for the v1.0 release. 

Hashing allows for optimal browser caching headers (ie. `public,max-age=31536000,immutable`).

Many frameworks perform this out-of-the-box Vite, Nextjs, etc

### Possible solutions

Hashing files: 
> For example, `imgUrl` will be `/img.png` during development, and become `/assets/img.2d8efhg.png` in the production build.

### Alternatives considered

Relying upon `ETag` and `Last-Modified` headers to handle invalidation

### Open Questions
- Currently, the css hashing doesn't seem to be consistent length, would we want to ensure a consistent length for easier pattern matching regexes? 
- Currently, the css hashing uses a format of `<name>-<hash>.<ext>` would we instead want `<name>.<hash>.<ext>`?
- Currently, the css hashing uses [a-zA-Z0-9] characters, I have seen various hashing algorithms used, just wanted to verify this was the one we were wanting to settle on.
- Would we want to move hashed assets to a single folder to make it even easier to apply caching headers? I believe Next.js does something like this.
- How does this impact dynamic imports?

### Help make it happen!

- [ ] I am willing to submit a PR to implement this change.
- [X] I am willing to submit a PR to implement this change, but would need some guidance.
- [ ] I am not willing to submit a PR to implement this change.

**Comment by matthewp** [comment]
Awesome @stramel, would love to talk about this on the RFC call tomorrow if you're available.

**Comment by stramel** [comment]
@matthewp I will be sure to hop on for the RFC tomorrow!

**Comment by FredKSchott** [comment]
RFC Call:
- Conceptually aligned with this
- would be best to scope this to just JS fingerprinting/hashing for now, while we work through other asset proposals like #492

**Comment by dtinth** [comment]
I had a problem where my website encountered a brief downtime due to some assets not being hashed.

1. I go to the website. The HTML page requests this file which gets cached:

   ```
   /_snowpack/pkg/@astrojs.renderer-svelte.client.v0.1.1.js
   ```

2. The file contains an imports which looks like this:

   ```js
   import {S as t, i as n, …} from "../../svelte.internal.v3.42.4-4009b5e1.js";
   ```

3. I deployed a new site version. The file contents changed to:

   ```js
   import {S as t, i as n, …} from "../../svelte.internal.v3.42.4-e1b2f404.js";
   ```

4. I visit the site. For some reason, the CDN returns the old version of `@astrojs.renderer-svelte.client.v0.1.1.js`.

5. But the file it points to, `svelte.internal.v3.42.4-4009b5e1.js` has been deleted already due to the deployment. This caused 404 and the JavaScript fails to load. I fixed this by purging the CDN cache.

So, +1 for hashing all files in the production build.
